spec/apibuilder-generator.json

```diff
@@ -21,6 +21,15 @@
     { "uri": "https://app.apibuilder.io/apicollective/apibuilder-spec/latest/service.json" }
   ],
 
+  "enums": {
+    "file_type": {
+      "description": "Allows generator authors to flag special files. It is up to the client (i.e. the cli) to decide how to interpret them.",
+      "values": [
+        { "name": "editable", "description": "The end user may edit these files - consider not overwriting these files when code is re-generated." }
```

Comment:
Two questions: 1) Would it make sense to rename this to file_flag and make the field in file a [file_flag], that way we can later support other flags that we may want to attach to a file 2) Rather than marking files editable or non-editable in the generator, have you looked at omitting files via the extended apibuilder-cli .apibuilder syntax, where you can specify which files that a generator outputs are actually downloaded (and thus overwritten)?

Reply:
The question being: With the ability to filter generated files via apibuilder-cli config, is this change really necessary?

Reply:
Re #1: I thought about that, but figured that it kind of boiled down to a bunch of booleans - and we know how Mike feels about booleans ;) But in reflecting upon it it probably makes sense to do it as a `[file_flag]` - that's essentially how files on a computer work. It ends up being a difference between `[editable, external, xxx]` vs a single enum of `editable_and_external_and_xxx` - agreed that the former is better.

Re #2: I did also think about this, but I think it's better if the generator author owns the flagging. That person knows what should and shouldn't be edited and is structuring the generated code in a certain way to accommodate for that. With your suggestion, the "client" of the generator would somehow need to know which files to edit vs not and would need to manually configure that. This would leave a gap of both knowledge sharing and manual execution. Maybe this could work if the author of the generator was exclusively writing that generator for him/herself - and thus can tightly couple the apibuilder-cli config to the generator. But outside of this, there would need to be some communication about which files are editable. I think a `file_flag` would accomplish that while also allowing the cli to automatically deal with it in a way that the client doesn't even need to know what's going on. It also allows for something like the diff feature I added in a latter commit on https://github.com/apicollective/apibuilder-cli/pull/33. Wdyt?

Reply:
Generally though you as the consumer/user of a generated file know whether or not a particular file was edited or not, and can take measures  that it doesn't get overridden (via excluding it from .apibuilder).

Reply:
Yeah, but this allows the generator to be opinionated, i.e. in the same way apibuilder says "this is how to do REST", the generator can say "this is how to implement (Play microservice|http4s|etc)" down to the level of what is generated and what is not.

Reply:
I guess the question is should apibuilder encourage users to edit generated files at all, or should this be the exception for hacks and work-arounds where it's purely the client's responsiblity that edits don't get overwritten.

Reply:
True. I think there is a core use case for generating server-side code - which would necessitate being able to edit code. Imagine you get all the routes/controllers/logging/etc for free from a generator - plus a service `trait` that must be implemented with the business logic. The generator can provide that implementation (because it needs to mix it in in the right places, with correct naming); this is what would be edited. I can imagine a short step from that model to being able to either a) generate new server code as the API changes (incremental change) or b) completely rewriting the service in a different framework (everything around the service layer changes). In both cases, the "client" could look at the diff between the old raw version of the editable file and the new version; this would make it obvious what changes need to happen - something that would be missed by simply ignoring changes to files that have been manually added.

In general, I think this is a new use case that we haven't strongly considered since we've been focused on client generation (which, I agree, should not be edited). As such, the considerations may be different than we have historically had.

Reply:
Taking the http4s server generator as an example, the generated code provide traits with an abstract handler function for each route. You'd extend that in a class in a separate non-generated file where you implement all route handler functions.

Reply:
Yeah, so same model except that the generator is giving you that implemented class file too. Where this becomes perhaps more useful is with files that are not compiled - think config files or Node implementation classes. For example, say you have a prod config file that extends/overrides a base application.conf. If the generator gave you a prod.conf that you could edit locally, you'd want to know if the generator is suggesting changes to that prod.conf. This proposal would allow those types of suggestions to be surfaced (i.e. "we added database configuration where there was none! Please add xxx to prod.conf or you'll be getting dev conf from the default application.conf!"). In the Node example, you'd easily see how your service implementation should change - which would otherwise be difficult to express because there is no trait/impl relationship that enables you to catch things at compile-time.

Reply:
FYI I've updated the other commits (in apibuilder-cli and apibuilder-generator) to use file_flag instead of file_type.

Reply:
Cool, thanks for the explanations!